File: protocol-generator/c/src/base58.c
/*
 * base58.c — Base58 (Bitcoin alphabet) encode/decode, hand-rolled.
 *
 * Used for peer_id formatting/parsing (V7 §1.2 / §7.3). Leading zero bytes map to
 * a leading '1' each (leading-zero preserving in both directions). Implemented by
 * byte-wise long division / multiplication (no bignum dep).
 *
 * Memory: *out is malloc'd; the caller frees with free().
 *
 * SPDX-License-Identifier: Apache-2.0
 */
#include "entity_core/protocol.h"

#include <stdlib.h>
#include <string.h>

static const char ALPHABET[] =
    "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz";

/* Reverse map: char -> value, -1 if not a base58 char. */
static int b58_index(char c)
{
    for (int i = 0; i < 58; i++) {
        if (ALPHABET[i] == c) {
            return i;
        }
    }
    return -1;
}

ec_status ec_base58_encode(const uint8_t *in, size_t in_len, char **out)
{
    if (!out) {
        return EC_ERR_BAD_INPUT;
    }
    *out = NULL;

    size_t zeros = 0;
    while (zeros < in_len && in[zeros] == 0) {
        zeros++;
    }

    /* Upper bound on the base58 digit count: ceil(in_len * log(256)/log(58)) + 1.
     * log(256)/log(58) ~= 1.365658; use 138/100 as a safe rational over-estimate. */
    size_t cap = in_len * 138 / 100 + 1;
    uint8_t *digits = calloc(cap > 0 ? cap : 1, 1);
    if (!digits) {
        return EC_ERR_OOM;
    }

    /* `length` = number of base58 digits currently held at the END of `digits`. */
    size_t length = 0;
    for (size_t b = 0; b < in_len; b++) {
        int carry = in[b];
        size_t k = cap; /* write index, walking backward */
        for (size_t processed = 0; processed < length || carry != 0; processed++) {
            k--;
            carry += 256 * digits[k];
            digits[k] = (uint8_t)(carry % 58);
            carry /= 58;
        }
        length = cap - k;
    }

    /* Most-significant digit starts at cap-length; no leading zero-digits remain. */
    size_t k = cap - length;

    size_t out_len = zeros + (cap - k);
    char *s = malloc(out_len + 1);
    if (!s) {
        free(digits);
        return EC_ERR_OOM;
    }
    size_t pos = 0;
    for (size_t z = 0; z < zeros; z++) {
        s[pos++] = '1';
    }
    for (; k < cap; k++) {
        s[pos++] = ALPHABET[digits[k]];
    }
    s[pos] = '\0';

    free(digits);
    *out = s;
    return EC_OK;
}

ec_status ec_base58_decode(const char *str, uint8_t **out, size_t *out_len)
{
    if (!str || !out || !out_len) {
        return EC_ERR_BAD_INPUT;
    }
    *out = NULL;
    *out_len = 0;

    size_t slen = strlen(str);
    size_t ones = 0;
    while (ones < slen && str[ones] == '1') {
        ones++;
    }

    /* Upper bound on the byte count: ceil(slen * log(58)/log(256)) + 1.
     * log(58)/log(256) ~= 0.7322; use 733/1000 as a safe over-estimate. */
    size_t cap = slen * 733 / 1000 + 1;
    uint8_t *bytes = calloc(cap > 0 ? cap : 1, 1);
    if (!bytes) {
        return EC_ERR_OOM;
    }

    /* `length` = number of base256 bytes currently held at the END of `bytes`. */
    size_t length = 0;
    for (size_t i = 0; i < slen; i++) {
        int d = b58_index(str[i]);
        if (d < 0) {
            free(bytes);
            return EC_ERR_BAD_INPUT;
        }
        int carry = d;
        size_t k2 = cap;
        for (size_t processed = 0; processed < length || carry != 0; processed++) {
            k2--;
            carry += 58 * bytes[k2];
            bytes[k2] = (uint8_t)(carry & 0xff);
            carry >>= 8;
        }
        length = cap - k2;
    }

    /* Most-significant byte starts at cap-length; no leading zero-bytes remain. */
    size_t k = cap - length;

    size_t body = cap - k;
    size_t total = ones + body;
    uint8_t *result = malloc(total > 0 ? total : 1);
    if (!result) {
        free(bytes);
        return EC_ERR_OOM;
    }
    memset(result, 0, ones); /* leading '1' -> 0x00 */
    if (body > 0) {
        memcpy(result + ones, bytes + k, body);
    }

    free(bytes);
    *out = result;
    *out_len = total;
    return EC_OK;
}

File: protocol-generator/zig/src/base58.zig
//! Base58 (Bitcoin alphabet) — hand-rolled, used for peer-id formatting.
//! Standard byte-array long-division; no bignum dependency (std-only, profile
//! [codec].base58_library = hand-rolled). Leading zero bytes map to leading '1'
//! characters. No-GC idiom: both directions return an owned slice (caller frees).

const std = @import("std");

pub const alphabet = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz";

pub const Error = error{ InvalidCharacter, OutOfMemory };

/// Base58-encode `input`. Returns an owned slice (caller frees).
pub fn encode(gpa: std.mem.Allocator, input: []const u8) Error![]u8 {
    const len = input.len;
    var zeros: usize = 0;
    while (zeros < len and input[zeros] == 0) zeros += 1;

    const size = (len * 138 / 100) + 1;
    const b58 = try gpa.alloc(u8, size);
    defer gpa.free(b58);
    @memset(b58, 0);

    var high: usize = size - 1;
    var i: usize = 0;
    while (i < len) : (i += 1) {
        var carry: usize = input[i];
        var j: usize = size - 1;
        while (j > high or carry != 0) {
            carry += 256 * @as(usize, b58[j]);
            b58[j] = @intCast(carry % 58);
            carry /= 58;
            if (j == 0) break;
            j -= 1;
        }
        high = j;
    }

    var start: usize = 0;
    while (start < size and b58[start] == 0) start += 1;

    var out: std.ArrayList(u8) = .empty;
    errdefer out.deinit(gpa);
    @memset(try out.addManyAsSlice(gpa, zeros), '1');
    var k: usize = start;
    while (k < size) : (k += 1) try out.append(gpa, alphabet[b58[k]]);
    return out.toOwnedSlice(gpa);
}

fn value(c: u8) i16 {
    return switch (c) {
        '1'...'9' => @as(i16, c - '1'),
        'A'...'H' => @as(i16, c - 'A' + 9),
        'J'...'N' => @as(i16, c - 'J' + 17),
        'P'...'Z' => @as(i16, c - 'P' + 22),
        'a'...'k' => @as(i16, c - 'a' + 33),
        'm'...'z' => @as(i16, c - 'm' + 44),
        else => -1,
    };
}

/// Base58-decode `s`. Returns an owned slice (caller frees).
pub fn decode(gpa: std.mem.Allocator, s: []const u8) Error![]u8 {
    const len = s.len;
    var ones: usize = 0;
    while (ones < len and s[ones] == '1') ones += 1;

    const size = (len * 733 / 1000) + 1; // log(58)/log(256) ≈ 0.733
    const b256 = try gpa.alloc(u8, size);
    defer gpa.free(b256);
    @memset(b256, 0);

    var high: usize = size - 1;
    var i: usize = 0;
    while (i < len) : (i += 1) {
        const d = value(s[i]);
        if (d < 0) return error.InvalidCharacter;
        var carry: usize = @intCast(d);
        var j: usize = size - 1;
        while (j > high or carry != 0) {
            carry += 58 * @as(usize, b256[j]);
            b256[j] = @truncate(carry);
            carry >>= 8;
            if (j == 0) break;
            j -= 1;
        }
        high = j;
    }

    var start: usize = 0;
    while (start < size and b256[start] == 0) start += 1;

    var out: std.ArrayList(u8) = .empty;
    errdefer out.deinit(gpa);
    @memset(try out.addManyAsSlice(gpa, ones), 0);
    try out.appendSlice(gpa, b256[start..]);
    return out.toOwnedSlice(gpa);
}

test "base58 round trip with leading zeros" {
    const gpa = std.testing.allocator;
    const cases = [_][]const u8{
        &.{0x00},
        &.{ 0x00, 0x00, 0x01, 0x02 },
        &.{ 0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 },
        "hello world",
    };
    for (cases) |c| {
        const enc = try encode(gpa, c);
        defer gpa.free(enc);
        const dec = try decode(gpa, enc);
        defer gpa.free(dec);
        try std.testing.expectEqualSlices(u8, c, dec);
    }
}
